Python/brc.c

```diff
@@ -0,0 +1,183 @@
+// Implementation of biased reference counting inter-thread queue.
+//
+// Biased reference counting maintains two refcount fields in each object:
+// ob_ref_local and ob_ref_shared. The true refcount is the sum of these two
+// fields. In some cases, when refcounting operations are split across threads,
+// the ob_ref_shared field can be negative (although the total refcount must
+// be at least zero). In this case, the thread that decremented the refcount
+// requests that the owning thread give up ownership and merge the refcount
+// fields. This file implements the mechanism for doing so.
+//
+// Each thread state maintains a queue of objects whose refcounts it should
+// merge. The thread states are stored in a per-interpreter hash table by
+// thread id. The hash table has a fixed size and uses a linked list to store
+// thread states within each bucket.
+//
+// The queueing thread uses the eval breaker mechanism to notify the owning
+// thread that it has objects to merge. Additionaly, all queued objects are
+// merged during GC.
+#include "Python.h"
+#include "pycore_object.h"      // _Py_ExplicitMergeRefcount
+#include "pycore_brc.h"         // struct _brc_thread_state
+#include "pycore_ceval.h"       // _Py_set_eval_breaker_bit
+#include "pycore_llist.h"       // struct llist_node
+#include "pycore_pystate.h"     // _PyThreadStateImpl
+
+#ifdef Py_GIL_DISABLED
+
+// Get the hashtable bucket for a given thread id.
+static struct _brc_bucket *
+get_bucket(PyInterpreterState *interp, uintptr_t tid)
+{
+    return &interp->brc.table[tid % _Py_BRC_NUM_BUCKETS];
+}
+
+// Find the thread state in a hash table bucket by thread id.
+static _PyThreadStateImpl *
+find_thread_state(struct _brc_bucket *bucket, uintptr_t thread_id)
+{
+    struct llist_node *node;
+    llist_for_each(node, &bucket->root) {
+        // Get the containing _PyThreadStateImpl from the linked-list node.
+        _PyThreadStateImpl *ts = llist_data(node, _PyThreadStateImpl,
+                                            brc.bucket_node);
+        if (ts->brc.tid == thread_id) {
+            return ts;
+        }
+    }
+    return NULL;
+}
+
+// Enqueue an object to be merged by the owning thread. This steals a
+// reference to the object.
+void
+_Py_brc_queue_object(PyObject *ob)
+{
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+
+    uintptr_t ob_tid = _Py_atomic_load_uintptr(&ob->ob_tid);
+    if (ob_tid == 0) {
+        // The owning thread may have concurrently decided to merge the
+        // refcount fields.
+        Py_DECREF(ob);
+        return;
+    }
+
+    struct _brc_bucket *bucket = get_bucket(interp, ob_tid);
+    PyMutex_Lock(&bucket->mutex);
+    _PyThreadStateImpl *tstate = find_thread_state(bucket, ob_tid);
+    if (tstate == NULL) {
+        // If we didn't find the owning thread then it must have already exited.
```

Comment:
...or the owning `tid` applies to a thread state in a different interpreter.  Are we still okay here in that case?

The one-to-many relationship between `ob_tid` and thread states makes me a little uneasy.  (The same goes for `brc.tid`, I suppose.)  From what I understand, `ob_ref_local` relates to the thread where `new_reference()` was last called on the object.  It is *not* tied to a specific thread state.  (This impacts functions like `_Py_ExplicitMergeRefcount()`.)

That's relevant to this PR since the BRC code here *is* tied to a specific thread state.  I'm just unclear if that's an actual problem or not.

Relatedly, something we've considered is storing a reference to the owning `PyInterpreterState` on each `PyTypeObject`.  Then you can get the owning interpreter from `ob_type`.  (It would be a little tricky for builtin static types though.)  Perhaps that would help here (if needed)?

Reply:
> ...or the owning tid applies to a thread state in a different interpreter. Are we still okay here in that case?

This doesn't make that case any better or worse. The thread might be decref'd in a different interpreter than it was created in, but it's the same interpreter as the `Py_DECREF()` call.

Allocating an object in one interpreter and deallocating in another is problematic both in the default build and in the free-threaded build. We have some tests like `SinglephaseInitTests.test_basic_multiple_interpreters_deleted_no_reset`, but they only test really basic objects. Once you have non-trivial Python destructors you will encounter segfaults.

> Relatedly, something we've considered is storing a reference to the owning PyInterpreterState on each PyTypeObject.

That might help debugging, but I don't think it's necessary for this PR.

Reply:
We definitely shouldn't decref an object in one interpreter that belongs to another interpreter.  It's just that in the owning interpreter we may incref the object before sending it to another interpreter, just to make sure it stays alive.  In that scenario, the owning interpreter must still be where the decref happens.  For example, the other interpreter can release its reference by queueing up the decref as a "pending call" on the owning interpreter.

Between what you said and me having thought about it a little more, I think the `ob_tid` one-to-may relationship isn't a problem here.  I don't know if it might impact any other code.